File: src/Application/WorkItems/WorkItemService.cs
using Application.Abstractions;
using Application.Common.Exceptions;
using Application.Common.Models;
using Domain.WorkItems;

namespace Application.WorkItems;

public sealed class WorkItemService(
    IWorkItemRepository workItemRepository,
    IUnitOfWork unitOfWork,
    IClock clock)
{
    public async Task<WorkItemDto> CreateAsync(CreateWorkItemRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        WorkItem workItem = WorkItem.Create(request.Title, request.Description, clock.UtcNow);

        await workItemRepository.AddAsync(workItem, cancellationToken);
        await unitOfWork.SaveChangesAsync(cancellationToken);

        return workItem.ToDto();
    }

    public async Task<WorkItemDto> GetAsync(Guid id, CancellationToken cancellationToken)
    {
        WorkItem? workItem = await workItemRepository.GetByIdAsync(id, cancellationToken);
        return workItem is null
            ? throw new NotFoundException($"Work item '{id}' was not found.")
            : workItem.ToDto();
    }

    public async Task<WorkItemDto> UpdateAsync(Guid id, UpdateWorkItemRequest request, CancellationToken cancellationToken)
    {
        ArgumentNullException.ThrowIfNull(request);

        WorkItem? workItem = await workItemRepository.GetByIdAsync(id, cancellationToken);
        if (workItem is null)
        {
            throw new NotFoundException($"Work item '{id}' was not found.");
        }

        DateTimeOffset now = clock.UtcNow;
        workItem.Rename(request.Title, now);
        workItem.UpdateDescription(request.Description, now);

        switch (request.Status)
        {
            case WorkItemStatus.Active:
                workItem.Activate(now);
                break;
            case WorkItemStatus.Completed:
                workItem.Complete(now);
                break;
        }

        await unitOfWork.SaveChangesAsync(cancellationToken);
        return workItem.ToDto();
    }

    public async Task<PagedResult<WorkItemDto>> ListAsync(int skip, int take, CancellationToken cancellationToken)
    {
        if (skip < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(skip), "Skip must be zero or greater.");
        }

        if (take <= 0 || take > 200)
        {
            throw new ArgumentOutOfRangeException(nameof(take), "Take must be between 1 and 200.");
        }

        IReadOnlyList<WorkItemDto> items = (await workItemRepository.ListAsync(skip, take, cancellationToken))
            .Select(workItem => workItem.ToDto())
            .ToArray();

        return new PagedResult<WorkItemDto>(items, skip, take);
    }
}

File: frontend/src/app/work-items/work-item.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  WorkItemDto,
  PagedResponse,
  CreateWorkItemRequest,
  UpdateWorkItemRequest,
} from './work-item.model';

@Injectable({ providedIn: 'root' })
export class WorkItemService {
  private readonly http = inject(HttpClient);
  private readonly base = '/api/v1/work-items';

  list(skip = 0, take = 20): Observable<PagedResponse<WorkItemDto>> {
    const params = new HttpParams().set('skip', skip).set('take', take);
    return this.http.get<PagedResponse<WorkItemDto>>(this.base, { params });
  }

  get(id: string): Observable<WorkItemDto> {
    return this.http.get<WorkItemDto>(`${this.base}/${id}`);
  }

  create(request: CreateWorkItemRequest): Observable<WorkItemDto> {
    return this.http.post<WorkItemDto>(this.base, request);
  }

  update(id: string, request: UpdateWorkItemRequest): Observable<WorkItemDto> {
    return this.http.put<WorkItemDto>(`${this.base}/${id}`, request);
  }
}
